### PR TITLE
Create specified dir if needed during save_mlir and fix vulkan device fetching without URI/ID

### PR DESCRIPTION
**Pull request by monorimet**
shark_importer's `save_mlir` fails if the directory doesn't exist. Create it if this is the case.

Soft fail a100 SD CI as image similarity check is expected to fail at the moment.

### CODE REVIEW

**Comment by monorimet** [comment]
@stellaraccident this issue started to show up after the latest torch-mlir pre-release (runner installed 20231121.1029) and ran into the following error:
https://github.com/nod-ai/SHARK/actions/runs/6947408869/job/18901119754#step:11:944
```
  File "/data/anush/actions-runner/_work/SHARK/SHARK/shark/torch_mlir_utils.py", line 77, in get_torch_mlir_module

    mlir_module = torch_mlir.compile(

                  ^^^^^^^^^^^^^^^^^^^

  File "/data/anush/actions-runner/_work/SHARK/SHARK/shark.venv/lib/python3.11/site-packages/torch_mlir/__init__.py", line 419, in compile

    class_annotator.exportNone(scripted._c._type())

TypeError: exportNone(): incompatible function arguments. The following argument types are supported:

    1. (self: torch_mlir._mlir_libs._jit_ir_importer.ClassAnnotator, arg0: c10::ClassType) -> None



Invoked with: ClassAnnotator {

}

, __torch__.torch.fx.graph_module._lambda

Error: Process completed with exit code 1.
```

I saw there were some significant changes in these parts within the past few days, does this ring any bells? I don't want to pull you away from anything so if it's a non-trivial fix I can dig into it

**Comment by stellaraccident** [comment]
> @stellaraccident this issue started to show up after the latest torch-mlir pre-release (runner installed 20231121.1029) and ran into the following error: https://github.com/nod-ai/SHARK/actions/runs/6947408869/job/18901119754#step:11:944
> 
> ```
>   File "/data/anush/actions-runner/_work/SHARK/SHARK/shark/torch_mlir_utils.py", line 77, in get_torch_mlir_module
> 
>     mlir_module = torch_mlir.compile(
> 
>                   ^^^^^^^^^^^^^^^^^^^
> 
>   File "/data/anush/actions-runner/_work/SHARK/SHARK/shark.venv/lib/python3.11/site-packages/torch_mlir/__init__.py", line 419, in compile
> 
>     class_annotator.exportNone(scripted._c._type())
> 
> TypeError: exportNone(): incompatible function arguments. The following argument types are supported:
> 
>     1. (self: torch_mlir._mlir_libs._jit_ir_importer.ClassAnnotator, arg0: c10::ClassType) -> None
> 
> 
> 
> Invoked with: ClassAnnotator {
> 
> }
> 
> , __torch__.torch.fx.graph_module._lambda
> 
> Error: Process completed with exit code 1.
> ```
> 
> I saw there was some refactoring in torch-mlir within the past few days, could this be related? I don't want to pull you away from anything so if it's a non-trivial fix I can dig into it.

The refactorings should not have had a visible impact, but this code path is next to impossible to support. This seems like one of our old issues with RTTI being busted across C++ boundaries that was maybe triggered because some plank got moved and was holding the world up. It is probably a trivial fix and a non-trivial analysis that will require looking closely at how this is getting used.

Either that or I'm reading the error wrong and this is more of a run of the mill thing. I can give you some advice on things that I usually try to isolate this stuff, but it is all in the category of trying to identify the issue.

**Comment by monorimet** [comment]
> > @stellaraccident this issue started to show up after the latest torch-mlir pre-release (runner installed 20231121.1029) and ran into the following error: https://github.com/nod-ai/SHARK/actions/runs/6947408869/job/18901119754#step:11:944
> > ```
> >   File "/data/anush/actions-runner/_work/SHARK/SHARK/shark/torch_mlir_utils.py", line 77, in get_torch_mlir_module
> > 
> >     mlir_module = torch_mlir.compile(
> > 
> >                   ^^^^^^^^^^^^^^^^^^^
> > 
> >   File "/data/anush/actions-runner/_work/SHARK/SHARK/shark.venv/lib/python3.11/site-packages/torch_mlir/__init__.py", line 419, in compile
> > 
> >     class_annotator.exportNone(scripted._c._type())
> > 
> > TypeError: exportNone(): incompatible function arguments. The following argument types are supported:
> > 
> >     1. (self: torch_mlir._mlir_libs._jit_ir_importer.ClassAnnotator, arg0: c10::ClassType) -> None
> > 
> > 
> > 
> > Invoked with: ClassAnnotator {
> > 
> > }
> > 
> > , __torch__.torch.fx.graph_module._lambda
> > 
> > Error: Process completed with exit code 1.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I saw there was some refactoring in torch-mlir within the past few days, could this be related? I don't want to pull you away from anything so if it's a non-trivial fix I can dig into it.
> 
> The refactorings should not have had a visible impact, but this code path is next to impossible to support. This seems like one of our old issues with RTTI being busted across C++ boundaries that was maybe triggered because some plank got moved and was holding the world up. It is probably a trivial fix and a non-trivial analysis that will require looking closely at how this is getting used.
> 
> Either that or I'm reading the error wrong and this is more of a run of the mill thing. I can give you some advice on things that I usually try to isolate this stuff, but it is all in the category of trying to identify the issue.

Thanks. That answers my question. I'll keep digging into it.

File: shark/shark_importer.py [hunk 1]
```diff
@@ -809,6 +809,8 @@ def save_mlir(
         dir = os.path.join(".", "shark_tmp")
     mlir_path = os.path.join(dir, model_name_mlir)
     print(f"saving {model_name_mlir} to {dir}")
+    if not os.path.exists(dir):
```
Comment (dan-garvey) [inline shark/shark_importer.py:812]: you can use os.mkdirs here because you want to make all folders in dir that don't exist

**Review by dan-garvey — DISMISSED** [review]

File: shark/shark_importer.py (hunk 1 above)
Comment (monorimet) [reply shark/shark_importer.py:812]: Fixed. PTAL

**Review by gpetters94 — APPROVED** [review]
LGTM

**Pull request merged**